In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from imblearn.over_sampling import SMOTE

In [3]:
district=pd.read_csv('district data.csv')
cardio=pd.read_csv('cardio_train.csv',sep=';')


In [4]:
cardio=cardio.drop('id',axis=1)

In [5]:
district

district  population  increase  sex ratio  literacy  \
0                   Nagaon     2823768     22.00        962     72.37   
1                   Dhubri     1949258     24.44        953     58.34   
2                 Sonitpur     1924110     15.55        956     67.34   
3                   Cachar     1736617     20.19        959     79.34   
4                  Barpeta     1693622     21.43        953     63.81   
5                   Kamrup     1517542     15.69        949     75.55   
6                 Tinsukia     1327929     15.47        952     69.66   
7                Dibrugarh     1326335     11.92        961     76.05   
8      Kamrup Metropolitan     1253938     18.34        936     88.71   
9                Karimganj     1228686     21.90        963     78.22   
10               Sivasagar     1151050      9.44        954     80.41   
11                  Jorhat     1092256      9.31        962     82.15   
12                Golaghat     1066888     12.75        964     77.43   
13               Lakhimpur     1042137     17.22        968     77.20   
14                Goalpara     1008183     22.64        964     67.37   
15                Morigaon      957423     23.34        967     68.03   
16      Karbi Anglong East      956313     17.58        951     69.25   
17                   Baksa      950075     10.74        974     69.25   
18                 Darrang      928500     22.19        954     63.08   
19               Kokrajhar      887142      5.21        959     65.22   
20                Udalguri      831668      9.61        973     65.41   
21                 Nalbari      771639     11.99        949     78.63   
22              Bongaigaon      738804     20.59        966     69.74   
23                 Dhemaji      686133     19.97        953     72.70   
24              Hailakandi      659296     21.45        951     74.33   
25                 Chirang      482162     11.34        969     63.55   
26              Dima Hasao      214102     13.84        932     77.54   
27               Charaideo      505384       NaN        955     71.00   
28                  Majuli      167304       NaN        955     61.34   
29               Biswanath      612491       NaN        968     88.66   
30                 Darrang     1225620       NaN        954     63.08   
31                   Hojai      931218       NaN        953     90.66   
32      Karbi Anglong west     1262334       NaN        954     69.25   
33  South Samara Maakachar      555114       NaN        968     50.76   

    density  latitude  longitude  
0       711   26.3480    92.6838  
1       896   26.0207    89.9743  
2       370   26.6770    92.6984  
3       459   24.7821    92.8570  
4       742   26.3304    91.0041  
5       489   25.9411    91.2891  
6       350   27.4886    95.3558  
7       392   27.4705    94.9125  
8      1313   26.1701    91.9490  
9       679   24.8649    92.3592  
10      431   26.9826    94.6425  
11      383   26.7509    94.2037  
12      305   26.5239    93.9623  
13      458   27.2064    94.1514  
14      553   26.1641    90.6252  
15      617   26.2529    92.3370  
16       92   26.1861    93.5813  
17      387   26.6935    91.5984  
18      586   26.4512    92.1058  
19      269   26.4014    90.2667  
20      413   26.7460    92.0959  
21      733   26.4446    91.4411  
22      676   26.5030    90.5536  
23      212   27.4811    94.5573  
24      497   24.6809    92.5648  
25      251   26.5387    90.6939  
26       44   25.3478    93.0176  
27      366   27.0373    95.0182  
28      490   27.0016    94.2243  
29      332   26.8518    93.3389  
30      586   26.4512    92.1058  
31      651   26.0017    92.8477  
32       88   25.9273    92.3814  
33      980   25.8048    89.8879

In [6]:
district['population'].sum()

36465041

##### Sex ratio is defined as the number of females per 1000 males in a given population
To calculate exact count of male and femail in each district:-
1)calculate of fraction of male and female population; 
2)multiply male and female fraction with total population of district 

district['female_fraction']=district['sex ratio']/(district['sex ratio']+1000)

In [7]:
district['female_fraction']=district['sex ratio']/(district['sex ratio']+1000)
district['male_fraction']=1-district['female_fraction']

In [8]:
district['female_count']=district['population']*district['female_fraction']

In [9]:
district['male_count']=district['population']*district['male_fraction']

In [10]:
district['male_count']=district['male_count'].astype(int)
district['female_count']=district['female_count'].astype(int)

In [11]:
district

district  population  increase  sex ratio  literacy  \
0                   Nagaon     2823768     22.00        962     72.37   
1                   Dhubri     1949258     24.44        953     58.34   
2                 Sonitpur     1924110     15.55        956     67.34   
3                   Cachar     1736617     20.19        959     79.34   
4                  Barpeta     1693622     21.43        953     63.81   
5                   Kamrup     1517542     15.69        949     75.55   
6                 Tinsukia     1327929     15.47        952     69.66   
7                Dibrugarh     1326335     11.92        961     76.05   
8      Kamrup Metropolitan     1253938     18.34        936     88.71   
9                Karimganj     1228686     21.90        963     78.22   
10               Sivasagar     1151050      9.44        954     80.41   
11                  Jorhat     1092256      9.31        962     82.15   
12                Golaghat     1066888     12.75        964     77.43   
13               Lakhimpur     1042137     17.22        968     77.20   
14                Goalpara     1008183     22.64        964     67.37   
15                Morigaon      957423     23.34        967     68.03   
16      Karbi Anglong East      956313     17.58        951     69.25   
17                   Baksa      950075     10.74        974     69.25   
18                 Darrang      928500     22.19        954     63.08   
19               Kokrajhar      887142      5.21        959     65.22   
20                Udalguri      831668      9.61        973     65.41   
21                 Nalbari      771639     11.99        949     78.63   
22              Bongaigaon      738804     20.59        966     69.74   
23                 Dhemaji      686133     19.97        953     72.70   
24              Hailakandi      659296     21.45        951     74.33   
25                 Chirang      482162     11.34        969     63.55   
26              Dima Hasao      214102     13.84        932     77.54   
27               Charaideo      505384       NaN        955     71.00   
28                  Majuli      167304       NaN        955     61.34   
29               Biswanath      612491       NaN        968     88.66   
30                 Darrang     1225620       NaN        954     63.08   
31                   Hojai      931218       NaN        953     90.66   
32      Karbi Anglong west     1262334       NaN        954     69.25   
33  South Samara Maakachar      555114       NaN        968     50.76   

    density  latitude  longitude  female_fraction  male_fraction  \
0       711   26.3480    92.6838         0.490316       0.509684   
1       896   26.0207    89.9743         0.487967       0.512033   
2       370   26.6770    92.6984         0.488753       0.511247   
3       459   24.7821    92.8570         0.489535       0.510465   
4       742   26.3304    91.0041         0.487967       0.512033   
5       489   25.9411    91.2891         0.486916       0.513084   
6       350   27.4886    95.3558         0.487705       0.512295   
7       392   27.4705    94.9125         0.490056       0.509944   
8      1313   26.1701    91.9490         0.483471       0.516529   
9       679   24.8649    92.3592         0.490576       0.509424   
10      431   26.9826    94.6425         0.488229       0.511771   
11      383   26.7509    94.2037         0.490316       0.509684   
12      305   26.5239    93.9623         0.490835       0.509165   
13      458   27.2064    94.1514         0.491870       0.508130   
14      553   26.1641    90.6252         0.490835       0.509165   
15      617   26.2529    92.3370         0.491612       0.508388   
16       92   26.1861    93.5813         0.487442       0.512558   
17      387   26.6935    91.5984         0.493414       0.506586   
18      586   26.4512    92.1058         0.488229       0.511771   
19      269   26.4014    90.2667         0.489535       0.510465   
20      413   26.7460    92.0959  

# Up this 
## We have population of male and female of every district

In [12]:
cardio

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
0      18393       2     168    62.0    110     80            1     1      0   
1      20228       1     156    85.0    140     90            3     1      0   
2      18857       1     165    64.0    130     70            3     1      0   
3      17623       2     169    82.0    150    100            1     1      0   
4      17474       1     156    56.0    100     60            1     1      0   
...      ...     ...     ...     ...    ...    ...          ...   ...    ...   
69995  19240       2     168    76.0    120     80            1     1      1   
69996  22601       1     158   126.0    140     90            2     2      0   
69997  19066       2     183   105.0    180     90            3     1      0   
69998  22431       1     163    72.0    135     80            1     2      0   
69999  20540       1     170    72.0    120     80            2     1      0   

       alco  active  cardio  
0         0       1       0  
1         0       1       1  
2         0       0       1  
3         0       1       1  
4         0       0       0  
...     ...     ...     ...  
69995     0       1       0  
69996     0       1       1  
69997     1       0       1  
69998     0       0       1  
69999     0       1       0  

[70000 rows x 12 columns]

### How gender is represented ?

### I m not able to decide in gender column, 1 represent male population or 2 represent male population

So, I perform small analysis
:- to compair the average hight, weight of the population denoted by 1 and 2
(assuming female are physically smaller then male)

In [13]:
cardio.groupby('gender')['height'].mean()

gender
1    161.355612
2    169.947895
Name: height, dtype: float64

In [14]:
cardio.groupby('gender')['weight'].mean()

gender
1    72.565605
2    77.257307
Name: weight, dtype: float64

In [15]:
cardio.groupby('gender')['smoke'].sum()

gender
1     813
2    5356
Name: smoke, dtype: int64

### Conclusion:-
1= female;
2= male.

In [16]:
cardio['gender'].value_counts()

1    45530
2    24470
Name: gender, dtype: int64

In [17]:
Female_datasets=cardio['gender'].value_counts()[1]
Male_datasets=cardio['gender'].value_counts()[2]

### We have a problem here

We don't have that much of male or female in our data set. Example  Nagaon have 1439229 no of males and 1384538 no of female

I have only 24470 rows of males in my cardio data set, so I can only assign 24470 data to any district. similarly, for female, I have only 45530 data.

To handle this:
lets do for one district first;
taking the case of district with highest population Nagaon.

In [18]:
district[district['district']=='Nagaon']

district  population  increase  sex ratio  literacy  density  latitude  \
0   Nagaon     2823768      22.0        962     72.37      711    26.348   

   longitude  female_fraction  male_fraction  female_count  male_count  
0    92.6838         0.490316       0.509684       1384538     1439229

Because Nagaon have the highest no of male count i will assign all 24470 data to it. But to maintain the Gender ratio i have to write one equation

=> (male_count) *  (sex_ratio/1000)

=> (24470)*(962/1000) = 23540.14

means i can only asign 23540 female dataset to Nagaon to maintain gender ratio

### conclusion:-
Male assigned = 24470;
Female assigned = 23540

# Now for other district

can't assign Male assigned = 24470 Female assigned = 23540 to all the district, otherwise it will change the relative population density of all the district to 1

Because all the district have same number of population as well as male female ratio in that condition

### To handle this Situation:-

we can use population column of district data set

## Normalisation male_fraction & female_count column of every district

Lets assume total male_fraction & female_count of Nagaon district(with highest population) as 1

then we can calculate the normalized population of other district

In [19]:
district['normalized_male_count']=district['male_count']/district['male_count'].max()
district['normalized_female_count']=district['female_count']/district['female_count'].max()

In [20]:
district

district  population  increase  sex ratio  literacy  \
0                   Nagaon     2823768     22.00        962     72.37   
1                   Dhubri     1949258     24.44        953     58.34   
2                 Sonitpur     1924110     15.55        956     67.34   
3                   Cachar     1736617     20.19        959     79.34   
4                  Barpeta     1693622     21.43        953     63.81   
5                   Kamrup     1517542     15.69        949     75.55   
6                 Tinsukia     1327929     15.47        952     69.66   
7                Dibrugarh     1326335     11.92        961     76.05   
8      Kamrup Metropolitan     1253938     18.34        936     88.71   
9                Karimganj     1228686     21.90        963     78.22   
10               Sivasagar     1151050      9.44        954     80.41   
11                  Jorhat     1092256      9.31        962     82.15   
12                Golaghat     1066888     12.75        964     77.43   
13               Lakhimpur     1042137     17.22        968     77.20   
14                Goalpara     1008183     22.64        964     67.37   
15                Morigaon      957423     23.34        967     68.03   
16      Karbi Anglong East      956313     17.58        951     69.25   
17                   Baksa      950075     10.74        974     69.25   
18                 Darrang      928500     22.19        954     63.08   
19               Kokrajhar      887142      5.21        959     65.22   
20                Udalguri      831668      9.61        973     65.41   
21                 Nalbari      771639     11.99        949     78.63   
22              Bongaigaon      738804     20.59        966     69.74   
23                 Dhemaji      686133     19.97        953     72.70   
24              Hailakandi      659296     21.45        951     74.33   
25                 Chirang      482162     11.34        969     63.55   
26              Dima Hasao      214102     13.84        932     77.54   
27               Charaideo      505384       NaN        955     71.00   
28                  Majuli      167304       NaN        955     61.34   
29               Biswanath      612491       NaN        968     88.66   
30                 Darrang     1225620       NaN        954     63.08   
31                   Hojai      931218       NaN        953     90.66   
32      Karbi Anglong west     1262334       NaN        954     69.25   
33  South Samara Maakachar      555114       NaN        968     50.76   

    density  latitude  longitude  female_fraction  male_fraction  \
0       711   26.3480    92.6838         0.490316       0.509684   
1       896   26.0207    89.9743         0.487967       0.512033   
2       370   26.6770    92.6984         0.488753       0.511247   
3       459   24.7821    92.8570         0.489535       0.510465   
4       742   26.3304    91.0041         0.487967       0.512033   
5       489   25.9411    91.2891         0.486916       0.513084   
6       350   27.4886    95.3558         0.487705       0.512295   
7       392   27.4705    94.9125         0.490056       0.509944   
8      1313   26.1701    91.9490         0.483471       0.516529   
9       679   24.8649    92.3592         0.490576       0.509424   
10      431   26.9826    94.6425         0.488229       0.511771   
11      383   26.7509    94.2037         0.490316       0.509684   
12      305   26.5239    93.9623         0.490835       0.509165   
13      458   27.2064    94.1514         0.491870       0.508130   
14      553   26.1641    90.6252         0.490835       0.509165   
15      617   26.2529    92.3370         0.491612       0.508388   
16       92   26.1861    93.5813         0.487442       0.512558   
17      387   26.6935    91.5984         0.493414       0.506586   
18      586   26.4512    92.1058         0.488229       0.511771   
19      269   26.4014    90.2667         0.489535       0.510465   
20      413   26.7460    92.0959  

### Now if we multiply normalized_male_count with (Male assigned = 24470)

we will get the total number of male assigned to a perticular district 

### Now if we multiply normalized_female_count with (Male assigned = 23540)

we will get the total number of female assigned to a perticular district 

by this way we are not disturbing the gender ratio as well as relative population density of any district

this way we can also preserve the Correlation inbetween the columns also

# Implimentation

In [21]:
district['male_assigned']=(district['normalized_male_count']*Male_datasets).astype(int)
district['female_assigned']=(district['normalized_female_count']*Male_datasets*(district['sex ratio']/1000)).astype(int)

In [22]:
district

district  population  increase  sex ratio  literacy  \
0                   Nagaon     2823768     22.00        962     72.37   
1                   Dhubri     1949258     24.44        953     58.34   
2                 Sonitpur     1924110     15.55        956     67.34   
3                   Cachar     1736617     20.19        959     79.34   
4                  Barpeta     1693622     21.43        953     63.81   
5                   Kamrup     1517542     15.69        949     75.55   
6                 Tinsukia     1327929     15.47        952     69.66   
7                Dibrugarh     1326335     11.92        961     76.05   
8      Kamrup Metropolitan     1253938     18.34        936     88.71   
9                Karimganj     1228686     21.90        963     78.22   
10               Sivasagar     1151050      9.44        954     80.41   
11                  Jorhat     1092256      9.31        962     82.15   
12                Golaghat     1066888     12.75        964     77.43   
13               Lakhimpur     1042137     17.22        968     77.20   
14                Goalpara     1008183     22.64        964     67.37   
15                Morigaon      957423     23.34        967     68.03   
16      Karbi Anglong East      956313     17.58        951     69.25   
17                   Baksa      950075     10.74        974     69.25   
18                 Darrang      928500     22.19        954     63.08   
19               Kokrajhar      887142      5.21        959     65.22   
20                Udalguri      831668      9.61        973     65.41   
21                 Nalbari      771639     11.99        949     78.63   
22              Bongaigaon      738804     20.59        966     69.74   
23                 Dhemaji      686133     19.97        953     72.70   
24              Hailakandi      659296     21.45        951     74.33   
25                 Chirang      482162     11.34        969     63.55   
26              Dima Hasao      214102     13.84        932     77.54   
27               Charaideo      505384       NaN        955     71.00   
28                  Majuli      167304       NaN        955     61.34   
29               Biswanath      612491       NaN        968     88.66   
30                 Darrang     1225620       NaN        954     63.08   
31                   Hojai      931218       NaN        953     90.66   
32      Karbi Anglong west     1262334       NaN        954     69.25   
33  South Samara Maakachar      555114       NaN        968     50.76   

    density  latitude  longitude  female_fraction  male_fraction  \
0       711   26.3480    92.6838         0.490316       0.509684   
1       896   26.0207    89.9743         0.487967       0.512033   
2       370   26.6770    92.6984         0.488753       0.511247   
3       459   24.7821    92.8570         0.489535       0.510465   
4       742   26.3304    91.0041         0.487967       0.512033   
5       489   25.9411    91.2891         0.486916       0.513084   
6       350   27.4886    95.3558         0.487705       0.512295   
7       392   27.4705    94.9125         0.490056       0.509944   
8      1313   26.1701    91.9490         0.483471       0.516529   
9       679   24.8649    92.3592         0.490576       0.509424   
10      431   26.9826    94.6425         0.488229       0.511771   
11      383   26.7509    94.2037         0.490316       0.509684   
12      305   26.5239    93.9623         0.490835       0.509165   
13      458   27.2064    94.1514         0.491870       0.508130   
14      553   26.1641    90.6252         0.490835       0.509165   
15      617   26.2529    92.3370         0.491612       0.508388   
16       92   26.1861    93.5813         0.487442       0.512558   
17      387   26.6935    91.5984         0.493414       0.506586   
18      586   26.4512    92.1058         0.488229       0.511771   
19      269   26.4014    90.2667         0.489535       0.510465   
20      413   26.7460    92.0959  

In [23]:
district['male_assigned'].sum()+district['female_assigned'].sum()

618589

In [24]:
district[['district','population','male_assigned','female_assigned']]

district  population  male_assigned  female_assigned
0                   Nagaon     2823768          24470            23540
1                   Dhubri     1949258          16969            16020
2                 Sonitpur     1924110          16724            15889
3                   Cachar     1736617          15072            14409
4                  Barpeta     1693622          14744            13919
5                   Kamrup     1517542          13238            12393
6                 Tinsukia     1327929          11566            10896
7                Dibrugarh     1326335          11499            11039
8      Kamrup Metropolitan     1253938          11012            10028
9                Karimganj     1228686          10642            10258
10               Sivasagar     1151050          10015             9475
11                  Jorhat     1092256           9465             9105
12                Golaghat     1066888           9235             8921
13               Lakhimpur     1042137           9003             8769
14                Goalpara     1008183           8727             8431
15                Morigaon      957423           8275             8044
16      Karbi Anglong East      956313           8333             7834
17                   Baksa      950075           8183             8069
18                 Darrang      928500           8079             7643
19               Kokrajhar      887142           7699             7360
20                Udalguri      831668           7166             7053
21                 Nalbari      771639           6731             6301
22              Bongaigaon      738804           6389             6197
23                 Dhemaji      686133           5973             5639
24              Hailakandi      659296           5745             5401
25                 Chirang      482162           4163             4063
26              Dima Hasao      214102           1884             1701
27               Charaideo      505384           4395             4166
28                  Majuli      167304           1454             1379
29               Biswanath      612491           5291             5154
30                 Darrang     1225620          10664            10089
31                   Hojai      931218           8106             7653
32      Karbi Anglong west     1262334          10983            10391
33  South Samara Maakachar      555114           4795             4671

final=pd.DataFrame()
final=final+cardio[cardio['gender']==1].sample(Male_datasets)+cardio[cardio['gender']==2].sample(Male_datasets*.962)

final= pd.concat([cardio[cardio['gender']==1].sample(Male_datasets), final], axis=1)
final= pd.concat([cardio[cardio['gender']==2].sample(int(Male_datasets*0.962)), final], axis=1)
final['district']='Nagaon'

In [25]:
smote = SMOTE(sampling_strategy='auto')
df = smote.fit_resample